fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/impl/screenhandler/client/NetworkingClient.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.screenhandler.client;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.gui.screen.Screen;
+import net.minecraft.client.gui.screen.ingame.HandledScreens;
+import net.minecraft.client.gui.screen.ingame.ScreenHandlerProvider;
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.screen.ScreenHandlerType;
+import net.minecraft.text.Text;
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.api.ClientModInitializer;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.api.network.ClientSidePacketRegistry;
+import net.fabricmc.fabric.impl.screenhandler.ExtendedScreenHandlerType;
+import net.fabricmc.fabric.impl.screenhandler.Packets;
+
+@Environment(EnvType.CLIENT)
+public final class NetworkingClient implements ClientModInitializer {
+	private static final Logger LOGGER = LogManager.getLogger();
+
+	@Override
+	public void onInitializeClient() {
+		ClientSidePacketRegistry.INSTANCE.register(Packets.OPEN_ID, (ctx, buf) -> openScreen(buf));
+	}
+
+	@SuppressWarnings({"rawtypes", "unchecked"})
+	private void openScreen(PacketByteBuf buf) {
+		int typeId = buf.readVarInt();
+		int syncId = buf.readVarInt();
+		Text title = buf.readText();
+		ScreenHandlerType<?> type = Registry.SCREEN_HANDLER.get(typeId);
```

Comment:
Hmm, is `Registry.get` safe in netty eventloop

Reply:
I think it's also done in vanilla?

Reply:
This one should be fine assuming registry isn't changed when the server/client play is active, not so much for client player though if there is a disconnection etc.

Reply:
The type also needs to be there for the deserialization of the extra data in `buf`.

Reply:
I mean the access to client player below is unsafe.

Reply:
![](https://user-images.githubusercontent.com/6596629/79682649-90ef5f80-822c-11ea-81be-8b400eb66b53.png)
Would this be a solution for the thread issue? (and do I have to manually `release` the slice or does the GC do its job)

Reply:
You create the retained slice outsdie of the runnable. Otherwise when the handler's method exits the buf is released and by the time you call `retainedSlice` on buf on main thread, the buf is gone

Reply:
In fact, it seems you can just call `buf.retain` before you call `execute` and just pass the `buf` to `openScreen`